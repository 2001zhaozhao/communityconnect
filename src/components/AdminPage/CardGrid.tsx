import React, { Component } from 'react';
import { connect } from 'react-redux';
import { CurrentPosition } from "community-connect";
import { OrganizationCard, SortBar } from "../../community-connect-ui/Common";
import SearchBar from '../Header/SearchBar';
import { getDistance } from '../../utils';

type Props = {
    currentPosition: CurrentPosition;
    resource?: {
        id: string
    }[];
    saveItem?: (resource: any) => void;
    handleFilter?: () => void;
};

type State = {
    dataSort: () => string[];
};

export class CardGrid extends Component<Props, State> {
    constructor(props: Props) {
        super(props)

        this.state = {
            dataSort: this.sortByAlphabet,
        }
    }

    getCloserResource = (a: string, b: string): number => {
        if (a && getDistance(a, this.props.currentPosition)
            > getDistance(b, this.props.currentPosition)) {
            return 1;
        }

        return -1;
    }

    getCloserName = (a: { name: string }, b: { name: string }): number => {
        if (a.name > b.name) return 1
        else if (a.name < b.name) return -1
        else return 0
    }

    sortByAlphabet = () => {
        return this.props.resource && this.props.resource.slice().sort(this.getCloserName);
    }

    sortByDistance = () => {
        return this.props.resource && this.props.resource.slice().sort(this.getCloserResource);
    }

    handleSortChange = (newSort: () => void) => {
        if (this.state.dataSort !== newSort)
            this.setState({
                // Set the dataSort variable to whichever sort function is chosen
                dataSort: newSort,
            });
    }

    render() {
        const sortOptions = [
            { key: 'Alphabetically', sort: this.sortByAlphabet, disabled: false }
            , { key: 'Distance', sort: this.sortByDistance, disabled: !this.props.currentPosition }
        ];

        // Render will be called every time this.props.data is updated, and every time handleSortChange
        // updates the this.state.dataSort variable.
        // this.state.dataSort() sorts data to feed into the OrganizationCards without modifying the
        // source of data
        const sortedData = this.state.dataSort();

        return (
            <>
                <SearchBar
                    type="text"
                    handleFilter={this.props.handleFilter}
                />
                <SortBar
                    onSortChange={this.handleSortChange}
                    sortOptions={sortOptions}
                />
                {
                    sortedData.map((resource, index) => (
                        <OrganizationCard
                            key={resource.id}
                            index={resource.id}
                            organization={resource}
                            currentPos={this.props.currentPosition}
                            saveItem={() => this.props.saveItem(resource)}
                            saveable={true}
                        />
                    ))
                }
            </>
        );
    }
};

function mapStateToProps(state: any, ownProps: any) {
    let res = [];
    //Not the most efficient logic, but it works. Will have to optimize this later
    for (let i = 0, len1 = state.searchedResource.length; i < len1; i++) {
        for (let j = 0, len2 = state.filteredResource.length; j < len2; j++) {
            if (state.searchedResource[i].id === state.filteredResource[j].id) {
                res.push(state.searchedResource[i])
            }
        }
    }

    return {
        resource: res
    }
}


export default connect(mapStateToProps)(CardGrid);
