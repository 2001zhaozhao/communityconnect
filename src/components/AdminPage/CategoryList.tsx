import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import * as resourceAction from '../../action/resourceDataAction';

import { Form, FormGroup, Input, Label } from "../../community-connect-ui/common";

type Props = {
    resource: any;
    actions: any;
    categories: any;
};

type State = {
    selectedCategory: any[];
};

export class CategoryList extends Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            selectedCategory: []
        }
        this.handleChange = this.handleChange.bind(this);
    }

    handleChange(selected: any) {
        let { selectedCategory } = this.state;
        let index = selectedCategory.indexOf(selected);
        index !== -1 ? selectedCategory.splice(index, 1) : selectedCategory.push(selected);
        let filteredResource = this.props.resource.filter((resource: any) => {
            return this.state.selectedCategory.some(searchCategory => (
                resource.categoryAutoSortScript.split(',').map((item: any) => ( 
                    item.trim()).includes(searchCategory)
                )
            ));
        });
        this.props.actions.filterByCategories(selectedCategory.length > 0 ? filteredResource : this.props.resource);
    }

    categoryMenuItems() {
        return this.props.categories.map((cat: any) =>
            <FormGroup key={cat}>
                <Input type="checkbox" key={cat} onChange={() => this.handleChange(cat)} />{cat}
            </FormGroup>);
    }
    render() {
        return (
            <Form>
                <Label>Filter by Category</Label>
                {this.categoryMenuItems()}
            </Form>
        )
    }
}

function mapStateToProps(state: any, ownProps: any) {
    return {
        categories: state.categories,
        resource: state.resource
    }
}
function mapDispatchToProps(dispatch: any) {
    return {
        actions: bindActionCreators(resourceAction, dispatch)
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(CategoryList);
