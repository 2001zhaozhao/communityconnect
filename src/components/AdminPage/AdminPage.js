import React from 'react';

import CategoryList from './CategoryList';
import CardGrid from './CardGrid';

const AdminPage = ({currentPosition}) => (
    <>
        <CategoryList/>
        <CardGrid
            currentPos={currentPosition}
        />
    </>
)

export default AdminPage;
