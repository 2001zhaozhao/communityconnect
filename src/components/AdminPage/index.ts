import { CardGrid } from "./CardGrid";
import { CategoryList } from "./CategoryList";

export {
    CardGrid,
    CategoryList
};
