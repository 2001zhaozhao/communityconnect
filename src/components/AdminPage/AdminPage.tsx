import React from 'react';
import { Coordinates } from "community-connect";

import { CardGrid, CategoryList } from "../AdminPage";

type Props = {
    currentPosition: Coordinates
    resource?: any;
    actions?: any;
    categories?: any;
};

export const AdminPage: React.FC<Props> = ({ currentPosition, resource, actions, categories }) => (
    <>
        <CategoryList
            resource={resource}
            actions={actions}
            categories={categories}
        />
        <CardGrid
            currentPosition={currentPosition}
        />
    </>
)
