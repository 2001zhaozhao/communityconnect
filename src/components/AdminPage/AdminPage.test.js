import React from 'react'
import {shallow} from 'enzyme/build'
import AdminPage from "./AdminPage";

it('render Admin Page ', () => {
  const adminPage =  shallow(<AdminPage />);
  expect(adminPage).toMatchSnapshot();
});