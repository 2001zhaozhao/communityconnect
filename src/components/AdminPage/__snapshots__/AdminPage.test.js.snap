// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Admin Page  1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdminPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Connect(CategoryList) />,
        <Connect(CardGrid)
          currentPos={undefined}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "currentPos": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "__emotion_base": "div",
      "__emotion_forwardProp": undefined,
      "__emotion_real": [Circular],
      "__emotion_styles": Array [
        "
    display: grid;
    grid-template-columns: auto;
    padding: ",
        "30px",
        ";
    gap: ",
        "30px",
        ";
    @media screen and (min-width: ",
        "768px",
        ") {
        grid-template-columns: 200px auto;
    }
",
      ],
      "defaultProps": undefined,
      "displayName": "Styled(div)",
      "render": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Connect(CategoryList) />,
          <Connect(CardGrid)
            currentPos={undefined}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "currentPos": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "__emotion_base": "div",
        "__emotion_forwardProp": undefined,
        "__emotion_real": [Circular],
        "__emotion_styles": Array [
          "
    display: grid;
    grid-template-columns: auto;
    padding: ",
          "30px",
          ";
    gap: ",
          "30px",
          ";
    @media screen and (min-width: ",
          "768px",
          ") {
        grid-template-columns: 200px auto;
    }
",
        ],
        "defaultProps": undefined,
        "displayName": "Styled(div)",
        "render": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`render Admin Page 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdminPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Connect(CategoryList) />,
        <Connect(CardGrid)
          currentPos={undefined}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "currentPos": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "__emotion_base": "div",
      "__emotion_forwardProp": undefined,
      "__emotion_real": [Circular],
      "__emotion_styles": Array [
        "
    display: grid;
    grid-template-columns: auto;
    padding: ",
        "30px",
        ";
    gap: ",
        "30px",
        ";
    @media screen and (min-width: ",
        "768px",
        ") {
        grid-template-columns: 200px auto;
    }
",
      ],
      "defaultProps": undefined,
      "displayName": "Styled(div)",
      "render": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Connect(CategoryList) />,
          <Connect(CardGrid)
            currentPos={undefined}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "currentPos": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "__emotion_base": "div",
        "__emotion_forwardProp": undefined,
        "__emotion_real": [Circular],
        "__emotion_styles": Array [
          "
    display: grid;
    grid-template-columns: auto;
    padding: ",
          "30px",
          ";
    gap: ",
          "30px",
          ";
    @media screen and (min-width: ",
          "768px",
          ") {
        grid-template-columns: 200px auto;
    }
",
        ],
        "defaultProps": undefined,
        "displayName": "Styled(div)",
        "render": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdminPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Connect(CategoryList) />,
        <Connect(CardGrid)
          currentPos={undefined}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "currentPos": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "__emotion_base": "div",
      "__emotion_forwardProp": undefined,
      "__emotion_real": [Circular],
      "__emotion_styles": Array [
        "
    display: grid;
    grid-template-columns: auto;
    padding: ",
        "30px",
        ";
    gap: ",
        "30px",
        ";
    @media screen and (min-width: ",
        "768px",
        ") {
        grid-template-columns: 200px auto;
    }
",
      ],
      "defaultProps": undefined,
      "displayName": "Styled(div)",
      "render": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Connect(CategoryList) />,
          <Connect(CardGrid)
            currentPos={undefined}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "currentPos": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "__emotion_base": "div",
        "__emotion_forwardProp": undefined,
        "__emotion_real": [Circular],
        "__emotion_styles": Array [
          "
    display: grid;
    grid-template-columns: auto;
    padding: ",
          "30px",
          ";
    gap: ",
          "30px",
          ";
    @media screen and (min-width: ",
          "768px",
          ") {
        grid-template-columns: 200px auto;
    }
",
        ],
        "defaultProps": undefined,
        "displayName": "Styled(div)",
        "render": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
